groups/bsl/bslstl/bslstl_badoptionalaccess.t.cpp

```diff
@@ -0,0 +1,274 @@
+
+//=============================================================================
+//                              MAIN PROGRAM
+//-----------------------------------------------------------------------------
+
+int main(int argc, char *argv[])
+{
+
+    int  test                = argc > 1 ? atoi(argv[1]) : 0;
+    bool verbose             = argc > 2;
+//  bool veryVerbose         = argc > 3;
+//  bool veryVeryVerbose     = argc > 4;
+//  bool veryVeryVeryVerbose = argc > 5;
+
+    printf("TEST " __FILE__ " CASE %d\n", test);
+
+    switch (test) { case 0:
+#ifdef BDE_BUILD_TARGET_EXC
+      case 4: {
+        // --------------------------------------------------------------------
+        // 'bsl::bad_optional_access' TYPEDEF
+        //
+        // Concerns:
+        //: 1 The 'bsl::bad_optional_access' is a typedef for
+        //:   'std::bad_optional_access' if 'std::bad_optional_access' is
+        //:    available.
+        //
+        // Plan:
+        //: 1 For concern 1, if we're using CPP17 library, check that
+        //:   'bsl::bad_optional_access' is the same type as
+        //:   'std::bad_optional_access' using 'bsl::is_same'.
+        //
+        // Testing:
+        //   typedef bsl::bad_optional_access
+        // --------------------------------------------------------------------
+        if (verbose) printf("\n'bsl::bad_optional_access' TYPEDEF"
+                            "\n==================================\n");
+
+#ifdef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+        ASSERT((bsl::is_same<bsl::bad_optional_access,
+                             std::bad_optional_access>::value));
+
+#endif //BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
+      } break;
+      case 3: {
+        // --------------------------------------------------------------------
+        // DEFAULT CONSTRUCTION AND 'what' METHOD
+        //
+        // Concerns:
+        //: 1 A 'bad_optional_access' object can be default constructed.
+        //: 2 Invoking a 'what' method on a 'bad_optional_access' object
+        //:   returns a "bad_optional_access" string.
+        //: 3 'what' method can be invoked on a const 'bad_optional_access'
+        //:   object
+        //
+        // Plan:
+        //: 1 For concern 1, default construct a 'bad_optional_access' object.
+        //: 2 For concern 2, invoke the 'what' method on a
+        //:   'bad_optional_access' object and check that the returned string
+        //:   is "bad_optional_access".
+        //: 3 For concern 3, in step 2, use a const qualified
+        //:   'bad_optional_access' object.
+        //
+        // Testing:
+        //   bad_optional_access();
+        //   const char *what() const;
+        // --------------------------------------------------------------------
+        if (verbose) printf("\nDEFAULT CONSTRUCTION AND 'what' METHOD"
+                            "\n======================================\n");
+#ifndef BSLS_LIBRARYFEATURES_HAS_CPP17_BASELINE_LIBRARY
```

Comment:
This test does not seem C++17 specific.

Reply:
In C++17, standard library provides the implementation of this exception type. Because C++17 bsl::optional for non allocator aware value types uses std::optional as the underlying type (which may throw std::bad_optional_type) , C++17 bsl::bad_optional_type is just an alias for std::bad_optional_type. It seems pointless to run the tests for the standard exception type.

Reply:
I see what you're saying.  That said, it's more complex to not test the type than to test it, and It would be a clear demonstration that our definition of the type matches the standard (because you can see them passing the same set of tests).

Reply:
Understood. I'll make the tests non C++ version specific.

Reply:
Actually, I can't make this one work in C++17 as the string returned from what() is implementation specific. I've added a comment instead.